test/integration/api_listener_integration_test.cc

```diff
@@ -0,0 +1,121 @@
+
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+
+using testing::ReturnRef;
+
+namespace Envoy {
+namespace {
+
+class ApiListenerIntegrationTest : public BaseIntegrationTest,
+                                   public testing::TestWithParam<Network::Address::IpVersion> {
+public:
+  ApiListenerIntegrationTest() : BaseIntegrationTest(GetParam(), config()) {
+    use_lds_ = false;
+    autonomous_upstream_ = true;
+  }
+
+  void SetUp() override { BaseIntegrationTest::initialize(); }
+
+  void TearDown() override {
+    test_server_.reset();
+    fake_upstreams_.clear();
+  }
+
+  static std::string config() {
+    return R"EOF(
+admin:
+  access_log_path: /dev/null
+  address:
+    socket_address:
+      address: 127.0.0.1
+      port_value: 0
+static_resources:
+  clusters:
+    name: cluster_0
+    load_assignment:
+      cluster_name: my_cds_cluster
+      endpoints:
+      - lb_endpoints:
+        - endpoint:
+            address:
+              socket_address:
+                address: 127.0.0.1
+                port_value: 0
+  listeners:
+  - name: default_listener
+    address:
+      socket_address:
+        address: 127.0.0.1
+        port_value: 0
+    filter_chains:
+      filters:
+  - name: api_listener
+    address:
+      socket_address:
+        address: 127.0.0.1
+        port_value: 1
+    api_listener:
+      api_listener:
+        "@type": type.googleapis.com/envoy.extensions.filters.network.http_connection_manager.v3.HttpConnectionManager
+        stat_prefix: hcm
+        route_config:
+          virtual_hosts:
+            name: integration
+            routes:
+              route:
+                cluster: cluster_0
+              match:
+                prefix: "/"
+            domains: "*"
+          name: route_config_0
+        http_filters:
+          - name: envoy.router
+            typed_config:
+              "@type": type.googleapis.com/envoy.extensions.filters.http.router.v3.Router
+
+      )EOF";
+  }
+
+  NiceMock<Http::MockStreamEncoder> stream_encoder_;
+};
+
+INSTANTIATE_TEST_SUITE_P(IpVersions, ApiListenerIntegrationTest,
+                         testing::ValuesIn(TestEnvironment::getIpVersionsForTest()),
+                         TestUtility::ipTestParamsToString);
+
+TEST_P(ApiListenerIntegrationTest, Basic) {
+  ConditionalInitializer test_ran;
+  test_server_->server().dispatcher().post([this, &test_ran]() -> void {
+    ASSERT_TRUE(test_server_->server().listenerManager().apiListener().has_value());
+    ASSERT_EQ("api_listener", test_server_->server().listenerManager().apiListener()->get().name());
+    ASSERT_NE(nullptr, test_server_->server().listenerManager().apiListener()->get().http());
+    auto* http_api_listener = test_server_->server().listenerManager().apiListener()->get().http();
+
+    ON_CALL(stream_encoder_, getStream()).WillByDefault(ReturnRef(stream_encoder_.stream_));
+    auto& stream_decoder = http_api_listener->newStream(stream_encoder_);
```

Comment:
Why are we using mocks here? Can't these be a real end to end tests with an usptream?

Reply:
I am mocking the client bits to set expectations on what the client will receive. I am not sure what real class would make sense to use here otherwise. Unless I create a small Stream/StreamEncoder class?

For instance, in the Envoy Mobile integration tests I will use the actual Http::DirectStream and the Http::DirectStreamCallbacks and assert on values received. But here is made sense to me to mock the client and have expectations.

Reply:
Ah OK, I see. Alright that seems fine, thanks.